operator/src/monitoring.rs

```diff
@@ -0,0 +1,591 @@
+                },
+                properties: ReportingTaskComponentProperties {
+                    metrics_endpoint_port: monitoring_info.metric_port.to_string(),
+                    instance_id: "${hostname(true)}".to_string(),
+                    metrics_strategy: "All Components".to_string(),
+                    send_jvm: "true".to_string(),
+                    ssl_context: None,
+                    client_auth: "No Authentication".to_string(),
+                },
+            }),
+            state: None,
+            status: None,
+        };
+
+        self.post(&monitoring_info.create_reporting_task_url(), &task)
+            .await
+    }
+
+    /// Delete a reporting_task with a certain `task_id`.
+    ///
+    /// # Arguments
+    /// * `monitoring_info` - Pod monitoring info including container ports, uid etc.
+    /// * `task_id` - The PrometheusReportingTask id.
+    /// * `task_revision` - The PrometheusReportingTask revision.
+    ///
+    pub async fn delete_reporting_task(
+        &self,
+        monitoring_info: &PodMonitoringInfo,
+        task_id: &str,
+        task_revision: &ReportingTaskRevision,
+    ) -> Result<Response, reqwest::Error> {
+        self.delete(&monitoring_info.delete_reporting_task_url(task_id, task_revision))
+            .await
+    }
+
+    /// Update a reporting_task to "Stopped" or "Running".
+    ///
+    /// # Arguments
+    /// * `monitoring_info` - Pod monitoring info including container ports, uid etc.
+    /// * `task_id` - The PrometheusReportingTask id.
+    /// * `task_revision` - The PrometheusReportingTask revision.
+    /// * `state` - The ReportingTaskState to update (STOPPED, RUNNING).
+    ///
+    pub async fn update_reporting_task_status(
+        &self,
+        monitoring_info: &PodMonitoringInfo,
+        task_id: &str,
+        task_revision: &ReportingTaskRevision,
+        state: ReportingTaskState,
+    ) -> Result<Response, reqwest::Error> {
+        let new_task = ReportingTask {
+            revision: task_revision.clone(),
+            disconnected_node_acknowledged: Some(false),
+            state: Some(state),
+            id: None,
+            component: None,
+            status: None,
+        };
+
+        self.put(
+            &monitoring_info.update_reporting_task_status_url(task_id),
+            &new_task,
+        )
+        .await
+    }
+
+    /// Check if the configured [`ReportingTask`] port (in NiFi) equals the container port (in the
+    /// pod). The metric port is the same for all pods / NiFi nodes.
+    ///
+    /// # Arguments
+    /// * `metrics_port` - The provided metrics container port.
+    /// * `task` - The [`ReportingTask`] to check
+    /// * `task_id` - The PrometheusReportingTask id.
+    ///
+    pub async fn match_metric_and_reporting_task_port(
+        &self,
+        metrics_port: u16,
+        task_component: &Option<ReportingTaskComponent>,
+        task_id: &str,
+    ) -> Result<(), NifiMonitoringError> {
```

Comment:
IMO this function is a normal predicate and can just return bool. The fact that ports don't match is expected here and not a functional error.

```suggestion
    pub async fn match_metric_and_reporting_task_port(
        &self,
        metrics_port: u16,
        task_component: &Option<ReportingTaskComponent>,
        task_id: &str,
    ) -> bool {
        task_component.map_or(false, |task| {
            task.properties.metrics_endpoint_port == metrics_port.to_string()
        })
```

Reply:
Done